repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/services/uriredaction/UriRedactionServiceKeystoneTest.groovy

```diff
@@ -0,0 +1,207 @@
+            }
+        }
+
+        and: "OpenTracingService has logged that keystone span was sent to tracer"
+        spanList.each {
+            def logLines = reposeLogSearch.searchByString("Span reported: $it")
+            assert logLines.size() == 1
+        }
+
+        and: "OpenTracingService has logged that span was sent to tracer"
+        def traceId = URLDecoder.decode(messageChain.handlings.get(0).request.headers.getFirstValue(TRACING_HEADER), "UTF-8")
+        def logLines = reposeLogSearch.searchByString("Span reported: $traceId")
+        logLines.size() == 1
+
+        and: "The sent trace doesn't have the un-redacted token in it"
+        assertUntilTrue({ assert fakeTracer.batches.collect({ it.spans }).flatten().collect({ it.operationName }).any({ it.contains('/v2.0/tokens/XXXXX') }) })
+    }
+
+    def "when a call is made that hits against a uri with multiple capture groups the uri should redacted"() {
+        given:
+        fakeIdentityV2Service.with {
+            client_token = UUID.randomUUID().toString()
+            client_tenantid = "mytenant"
+            client_tenantname = "mytenantname"
+            client_userid = "12345"
+        }
+
+        def spanList = []
+
+        when: "User passes a request through repose with valid token"
+        MessageChain messageChain = deproxy.makeRequest(url: reposeEndpoint + "/foo/bar/baz", method: 'GET',
+            headers: ['content-type': 'application/json', 'X-Auth-Token': fakeIdentityV2Service.client_token])
+
+        then: "The request should have reached the origin service"
+        messageChain.handlings.size() == 1
+
+        and: "keystone request contains tracing header"
+        messageChain.orphanedHandlings.each {
+            // first, check to make sure it's not the tracer request (since that's done asynchronously and may count in orphaned handlings)
+            if (it.request.path != "/?format=jaeger.thrift") {
+                assert it.request.headers.contains(TRACING_HEADER)
+                def traceId = URLDecoder.decode(it.request.headers.getFirstValue(TRACING_HEADER), "UTF-8")
+                spanList << traceId
+            }
+        }
+
+        and: "OpenTracingService has logged that keystone span was sent to tracer"
+        spanList.each {
+            def logLines = reposeLogSearch.searchByString("Span reported: $it")
+            assert logLines.size() == 1
+        }
+
+        and: "OpenTracingService has logged that span was sent to tracer"
+        def traceId = URLDecoder.decode(messageChain.handlings.get(0).request.headers.getFirstValue(TRACING_HEADER), "UTF-8")
+        def logLines = reposeLogSearch.searchByString("Span reported: $traceId")
+        logLines.size() == 1
+
+        and: "The sent trace doesn't have the un-redacted token in it"
+        assertUntilTrue({ assert fakeTracer.batches.collect({ it.spans}).flatten().collect({ it.operationName }).any({ it.contains('/XXXXX/bar/XXXXX') }) })
```

Comment:
Missing a space after `it.spans`.

Reply:
I ran the formatter on the file and then decided to flatten this into a `Unroll` for readability.